projects/igniteui-angular/src/lib/tabs/tabs.directive.ts

```diff
@@ -0,0 +1,318 @@
+        }
+
+        const tabs = this.items.toArray();
+
+        if (selected) {
+            const index = tabs.indexOf(tab);
+            this.selectedIndex = index;
+        } else {
+            if (tabs.every(t => !t.selected)) {
+                this.selectedIndex = -1;
+            }
+        }
+    }
+
+    /** @hidden */
+    protected getPreviousElement(): HTMLElement {
+        return this.previousSlide.panelComponent.nativeElement;
+    }
+
+    /** @hidden */
+    protected getCurrentElement(): HTMLElement {
+        return this.currentSlide.panelComponent.nativeElement;
+    }
+
+    /** @hidden */
+    protected scrollTabHeaderIntoView() {
+    }
+
+    private setAttributes() {
+        this.items.forEach(item => {
+            if (item.panelComponent && !item.headerComponent.nativeElement.getAttribute('id')) {
+                const id = this.getNextTabId();
+                const tabHeaderId = `${this.componentName}-header-${id}`;
+                const tabPanelId = `${this.componentName}-content-${id}`;
+
+                this.setHeaderAttribute(item, 'id', tabHeaderId);
+                this.setHeaderAttribute(item, 'aria-controls', tabPanelId);
+                this.setPanelAttribute(item, 'id', tabPanelId);
+                this.setPanelAttribute(item, 'aria-labelledby', tabHeaderId);
+            }
+        });
+    }
+
+    private setHeaderAttribute(item: IgxTabItemDirective, attrName: string, value: string) {
+        item.headerComponent.nativeElement.setAttribute(attrName, value);
+    }
+
+    private setPanelAttribute(item: IgxTabItemDirective, attrName: string, value: string) {
+        item.panelComponent.nativeElement.setAttribute(attrName, value);
+    }
+
+    private get hasPanels() {
+        return this.panels && this.panels.length;
+    }
+
+    private updateSelectedTabs(oldSelectedIndex: number, raiseEvent = true) {
+        if (!this.items) {
+            return;
+        }
+
+        const tabs = this.items.toArray();
+        let newTab: IgxTabItemDirective;
+        const oldTab = this.currentSlide;
+
+        // First select the new tab
+        if (this._selectedIndex >= 0 && this._selectedIndex < tabs.length) {
+            newTab = tabs[this._selectedIndex];
+            newTab.selected = true;
+        }
+        // Then unselect the other tabs
+        tabs.forEach((tab, i) => {
+            if (i !== this._selectedIndex) {
+                tab.selected = false;
+            }
+        });
+
+        if (this._selectedIndex !== oldSelectedIndex) {
+            this.scrollTabHeaderIntoView();
+            this.triggerPanelAnimations(oldSelectedIndex);
+
+            if (raiseEvent && newTab !== oldTab) {
+                this.selectedItemChange.emit({
+                    owner: this,
+                    newItem: newTab,
+                    oldItem: oldTab
+                });
+            }
+        }
+    }
+
+    private onItemChanges() {
+        const tabs = this.items.toArray();

```

Comment:
Seems to be using it just for accessing the length of the collection which QueryList already provides.